File: app/src/main/java/ai/anaha/signup/hlsplayer/adapters/TrackSelectionViewAdapter.java
package ai.anaha.signup.hlsplayer.adapters;

import static android.view.View.INVISIBLE;
import static android.view.View.VISIBLE;

import android.util.SparseArray;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.RecyclerView;

import com.google.android.exoplayer2.trackselection.DefaultTrackSelector;
import com.google.android.exoplayer2.trackselection.MappingTrackSelector;
import com.google.android.exoplayer2.ui.TrackSelectionView;
import com.google.android.exoplayer2.util.Assertions;
import com.google.android.exoplayer2.util.Util;

import java.util.ArrayList;
import java.util.List;

import ai.anaha.signup.hlsplayer.R;
import ai.anaha.signup.hlsplayer.fragments.TrackSelectionFragment;
import ai.anaha.signup.hlsplayer.hlsutils.PlayerSettingsDialog;

public final class TrackSelectionViewAdapter extends RecyclerView.Adapter<TrackSelectionViewAdapter.SubSettingViewHolder> {

    private final ArrayList<String> trackNames;
    private final PlayerSettingsDialog.OnCustomTrackSelectedListener onCustomTrackSelectedListener;
    private final ArrayList<TrackSelectionFragment.TrackInfo> trackInfos;
    private final TrackSelectionView.TrackSelectionListener listener;
    private final MappingTrackSelector.MappedTrackInfo mappedTrackInfo;
    private final boolean allowMultipleOverrides;
    private int selectedIndex = 0;
    private final SparseArray<DefaultTrackSelector.SelectionOverride> overrides;
    private boolean isDisabled = false;
    private static final int VIEW_AUTO = 522;
    private static final int VIEW_TRACK = 523;
    private View lastCheckedView;

    public TrackSelectionViewAdapter(MappingTrackSelector.MappedTrackInfo mappedTrackInfo,
                                     ArrayList<String> trackNames, ArrayList<TrackSelectionFragment.TrackInfo> trackInfos, boolean allowMultipleOverrides,
                                     List<DefaultTrackSelector.SelectionOverride> overrides, PlayerSettingsDialog.OnCustomTrackSelectedListener onCustomTrackSelectedListener,
                                     @Nullable TrackSelectionView.TrackSelectionListener listener) {
        this.mappedTrackInfo = mappedTrackInfo;
        this.trackNames = trackNames;
        this.trackInfos = trackInfos;
        this.onCustomTrackSelectedListener = onCustomTrackSelectedListener;
        this.overrides = new SparseArray<>();
        this.allowMultipleOverrides = allowMultipleOverrides;
        int maxOverrides = allowMultipleOverrides ? overrides.size() : Math.min(overrides.size(), 1);
        for (int i = 0; i < maxOverrides; i++) {
            DefaultTrackSelector.SelectionOverride override = overrides.get(i);
            this.overrides.put(override.groupIndex, override);
        }
        this.listener = listener;
    }

    /*public void updateSelectedIndex(float playbackSpeed) {
        int currentSpeedMultBy100 = Math.round(playbackSpeed * 100);
        int closestMatchIndex = 0;
        int closestMatchDifference = Integer.MAX_VALUE;
        for (int i = 0; i < trackNames.length; i++) {
            int difference = Math.abs(currentSpeedMultBy100 - trackNames[i]);
            if (difference < closestMatchDifference) {
                closestMatchIndex = i;
                closestMatchDifference = difference;
            }
        }
        selectedIndex = closestMatchIndex;
    }*/

    @Override
    public SubSettingViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View v =
                LayoutInflater.from(parent.getContext())
                        .inflate(
                                R.layout.custom_exo_styled_track_list_item, parent, /* attachToRoot= */ false);
        return new SubSettingViewHolder(v);
    }

    @Override
    public void onBindViewHolder(@NonNull SubSettingViewHolder holder, int position) {
        if (position < getItemCount()) {
            holder.textView.setText(trackNames.get(position));
        }
        if (getItemViewType(position) == VIEW_AUTO) {
            if (selectedIndex == position && overrides.get(0) == null) {
                holder.checkView.setVisibility(VISIBLE);
                lastCheckedView = holder.checkView;
            } else {
                holder.checkView.setVisibility(INVISIBLE);
            }
            /*((LinearLayout) holder.itemView).addView(LayoutInflater.from(holder.itemView.getContext())
                    .inflate(R.layout.exo_list_divider, (LinearLayout) holder.itemView, false));*/
            holder.itemView.setOnClickListener(
                    v -> {
                        if (position != selectedIndex) {
                            lastCheckedView.setVisibility(INVISIBLE);
                            holder.checkView.setVisibility(VISIBLE);
                            onDefaultViewClicked();
                            listener.onTrackSelectionChanged(isDisabled, getOverrides());
                            onCustomTrackSelectedListener.onTrackSelected();
                        }
                    });
        } else {
            holder.checkView.setTag(trackInfos.get(position));
            holder.itemView.setOnClickListener(
                    v -> {
                        if (position != selectedIndex) {
                            onTrackViewClicked(holder.checkView);
                            listener.onTrackSelectionChanged(isDisabled, getOverrides());
                            onCustomTrackSelectedListener.onTrackSelected();
                            lastCheckedView.setVisibility(INVISIBLE);
                            holder.checkView.setVisibility(VISIBLE);
                        }
                    });

            DefaultTrackSelector.SelectionOverride override = overrides.get(0);
            if (override != null) {
                TrackSelectionFragment.TrackInfo trackInfo = Assertions.checkNotNull(trackInfos.get(position));
                if (override.containsTrack(trackInfo.trackIndex)) {
                    selectedIndex = holder.getAdapterPosition();
                    holder.checkView.setVisibility(VISIBLE);
                    lastCheckedView = holder.checkView;
                } else
                    holder.checkView.setVisibility(INVISIBLE);
            } else {
                holder.checkView.setVisibility(INVISIBLE);
            }
        }

    }

    private void onDefaultViewClicked() {
        isDisabled = false;
        overrides.clear();
    }

    public List<DefaultTrackSelector.SelectionOverride> getOverrides() {
        List<DefaultTrackSelector.SelectionOverride> overrideList = new ArrayList<>(overrides.size());
        for (int i = 0; i < overrides.size(); i++) {
            overrideList.add(overrides.valueAt(i));
        }
        return overrideList;
    }

    private void onTrackViewClicked(View view) {
        isDisabled = false;
        TrackSelectionFragment.TrackInfo trackInfo = (TrackSelectionFragment.TrackInfo) Assertions.checkNotNull(view.getTag());
        int groupIndex = trackInfo.groupIndex;
        int trackIndex = trackInfo.trackIndex;
        DefaultTrackSelector.SelectionOverride override = overrides.get(groupIndex);
        Assertions.checkNotNull(mappedTrackInfo);
        if (override == null) {
            // Start new override.
            if (!allowMultipleOverrides && overrides.size() > 0) {
                // Removed other overrides if we don't allow multiple overrides.
                overrides.clear();
            }
            overrides.put(groupIndex, new DefaultTrackSelector.SelectionOverride(groupIndex, trackIndex));
        } else {
            // An existing override is being modified.
            int overrideLength = override.length;
            int[] overrideTracks = override.tracks;
            boolean isCurrentlySelected = view.getVisibility() == VISIBLE;
            if (isCurrentlySelected) {
                // Remove the track from the override.
                if (overrideLength == 1) {
                    // The last track is being removed, so the override becomes empty.
                    overrides.remove(groupIndex);
                } else {
                    int[] tracks = getTracksRemoving(overrideTracks, trackIndex);
                    overrides.put(groupIndex, new DefaultTrackSelector.SelectionOverride(groupIndex, tracks));
                }
            } else {
                // Replace existing track in override.
                overrides.put(groupIndex, new DefaultTrackSelector.SelectionOverride(groupIndex, trackIndex));
            }
        }
    }


    private static int[] getTracksRemoving(int[] tracks, int removedTrack) {
        int[] newTracks = new int[tracks.length - 1];
        int trackCount = 0;
        for (int track : tracks) {
            if (track != removedTrack) {
                newTracks[trackCount++] = track;
            }
        }
        return newTracks;
    }

    @Override
    public int getItemViewType(int position) {
        return (position == 0) ? VIEW_AUTO : VIEW_TRACK;
    }

    @Override
    public int getItemCount() {
        return trackNames.size();
    }

    protected static class SubSettingViewHolder extends RecyclerView.ViewHolder {

        public final TextView textView;
        public final View checkView;

        public SubSettingViewHolder(View itemView) {
            super(itemView);
            if (Util.SDK_INT < 26) {
                // Workaround for https://github.com/google/ExoPlayer/issues/9061.
                itemView.setFocusable(true);
            }
            textView = itemView.findViewById(com.google.android.exoplayer2.ui.R.id.exo_text);
            checkView = itemView.findViewById(com.google.android.exoplayer2.ui.R.id.exo_check);
        }
    }
}

File: HLSPlayer/src/main/java/ai/anaha/hlsplayer/adapters/TrackSelectionViewAdapter.kt
package ai.anaha.hlsplayer.adapters

import ai.anaha.hlsplayer.R
import ai.anaha.hlsplayer.fragments.TrackSelectionFragment
import ai.anaha.hlsplayer.hlsutils.PlayerSettingsDialog.OnCustomTrackSelectedListener
import android.util.SparseArray
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.google.android.exoplayer2.trackselection.DefaultTrackSelector.SelectionOverride
import com.google.android.exoplayer2.trackselection.MappingTrackSelector.MappedTrackInfo
import com.google.android.exoplayer2.ui.TrackSelectionView.TrackSelectionListener
import com.google.android.exoplayer2.util.Assertions
import com.google.android.exoplayer2.util.Util
import java.util.*

class TrackSelectionViewAdapter(
    private val mappedTrackInfo: MappedTrackInfo,
    private val trackNames: ArrayList<String>,
    private val trackInfos: ArrayList<TrackSelectionFragment.TrackInfo?>,
    private val allowMultipleOverrides: Boolean,
    overrides: List<SelectionOverride>,
    private val onCustomTrackSelectedListener: OnCustomTrackSelectedListener,
    listener: TrackSelectionListener?
) : RecyclerView.Adapter<TrackSelectionViewAdapter.SubSettingViewHolder>() {
    private val listener: TrackSelectionListener?
    private var selectedIndex = 0
    private val overrides: SparseArray<SelectionOverride?> = SparseArray()
    private var isDisabled = false
    private var lastCheckedView: View? = null
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): SubSettingViewHolder {
        val v = LayoutInflater.from(parent.context)
            .inflate(
                R.layout.custom_exo_styled_track_list_item, parent,  /* attachToRoot= */false
            )
        return SubSettingViewHolder(v)
    }

    override fun onBindViewHolder(holder: SubSettingViewHolder, position: Int) {
        if (position < itemCount) {
            holder.textView.text = trackNames[position]
        }
        if (getItemViewType(position) == VIEW_AUTO) {
            if (selectedIndex == position && overrides[0] == null) {
                holder.checkView.visibility = View.VISIBLE
                lastCheckedView = holder.checkView
            } else {
                holder.checkView.visibility = View.INVISIBLE
            }
            holder.itemView.setOnClickListener {
                if (position != selectedIndex) {
                    lastCheckedView!!.visibility = View.INVISIBLE
                    holder.checkView.visibility = View.VISIBLE
                    onDefaultViewClicked()
                    listener!!.onTrackSelectionChanged(isDisabled, getOverrides())
                    onCustomTrackSelectedListener.onTrackSelected()
                }
            }
        } else {
            holder.checkView.tag = trackInfos[position]
            holder.itemView.setOnClickListener {
                if (position != selectedIndex) {
                    onTrackViewClicked(holder.checkView)
                    listener!!.onTrackSelectionChanged(isDisabled, getOverrides())
                    onCustomTrackSelectedListener.onTrackSelected()
                    lastCheckedView!!.visibility = View.INVISIBLE
                    holder.checkView.visibility = View.VISIBLE
                }
            }
            val override = overrides[0]
            if (override != null) {
                val trackInfo = Assertions.checkNotNull(
                    trackInfos[position]
                )
                if (override.containsTrack(trackInfo.trackIndex)) {
                    selectedIndex = holder.absoluteAdapterPosition
                    holder.checkView.visibility = View.VISIBLE
                    lastCheckedView = holder.checkView
                } else holder.checkView.visibility = View.INVISIBLE
            } else {
                holder.checkView.visibility = View.INVISIBLE
            }
        }
    }

    private fun onDefaultViewClicked() {
        isDisabled = false
        overrides.clear()
    }

    private fun getOverrides(): MutableList<SelectionOverride> {
        val overrideList: MutableList<SelectionOverride> = mutableListOf()
        for (i in 0 until overrides.size()) {
            overrides.valueAt(i)?.let { overrideList.add(it) }
        }
        return overrideList
    }

    private fun onTrackViewClicked(view: View) {
        isDisabled = false
        val trackInfo = Assertions.checkNotNull(view.tag) as TrackSelectionFragment.TrackInfo
        val groupIndex = trackInfo.groupIndex
        val trackIndex = trackInfo.trackIndex
        val override = overrides[groupIndex]
        Assertions.checkNotNull(mappedTrackInfo)
        if (override == null) {
            // Start new override.
            if (!allowMultipleOverrides && overrides.size() > 0) {
                // Removed other overrides if we don't allow multiple overrides.
                overrides.clear()
            }
            overrides.put(groupIndex, SelectionOverride(groupIndex, trackIndex))
        } else {
            // An existing override is being modified.
            val overrideLength = override.length
            val overrideTracks = override.tracks
            val isCurrentlySelected = view.visibility == View.VISIBLE
            if (isCurrentlySelected) {
                // Remove the track from the override.
                if (overrideLength == 1) {
                    // The last track is being removed, so the override becomes empty.
                    overrides.remove(groupIndex)
                } else {
                    val tracks = getTracksRemoving(overrideTracks, trackIndex)
                    overrides.put(groupIndex, SelectionOverride(groupIndex, *tracks))
                }
            } else {
                // Replace existing track in override.
                overrides.put(groupIndex, SelectionOverride(groupIndex, trackIndex))
            }
        }
    }

    override fun getItemViewType(position: Int): Int {
        return if (position == 0) VIEW_AUTO else VIEW_TRACK
    }

    override fun getItemCount(): Int {
        return trackNames.size
    }

    class SubSettingViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val textView: TextView
        val checkView: View

        init {
            if (Util.SDK_INT < 26) {
                // Workaround for https://github.com/google/ExoPlayer/issues/9061.
                itemView.isFocusable = true
            }
            textView = itemView.findViewById(com.google.android.exoplayer2.ui.R.id.exo_text)
            checkView = itemView.findViewById(com.google.android.exoplayer2.ui.R.id.exo_check)
        }
    }

    companion object {
        private const val VIEW_AUTO = 522
        private const val VIEW_TRACK = 523
        private fun getTracksRemoving(tracks: IntArray, removedTrack: Int): IntArray {
            val newTracks = IntArray(tracks.size - 1)
            var trackCount = 0
            for (track in tracks) {
                if (track != removedTrack) {
                    newTracks[trackCount++] = track
                }
            }
            return newTracks
        }
    }

    init {
        val maxOverrides =
            if (allowMultipleOverrides) overrides.size else Math.min(overrides.size, 1)
        for (i in 0 until maxOverrides) {
            val override = overrides[i]
            this.overrides.put(override.groupIndex, override)
        }
        this.listener = listener
    }
}